app/src/main/java/io/apicurio/registry/config/RegistryConfigServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+import io.apicurio.registry.storage.RegistryStorage;
+import io.apicurio.registry.types.Current;
+import io.quarkus.scheduler.Scheduled;
+
+/**
+ * @author eric.wittmann@gmail.com
+ */
+@ApplicationScoped
+public class RegistryConfigServiceImpl implements RegistryConfigService {
+
+    private static final String DEFAULT_TENANT_ID = "___";
+
+    @Inject
+    Logger log;
+
+    @Inject
+    @Current
+    RegistryStorage storage;
+
+    @Inject
+    TenantContext tenantContext;
+
+    private Map<RegistryConfigProperty, Object> globalPropertyCache = new HashMap<>();
+    private Map<String, Map<RegistryConfigProperty, Object>> tenantPropertyCaches = new HashMap<>();
+    private Instant lastRefresh = null;
+
+    /**
+     * @see io.apicurio.registry.config.RegistryConfigService#get(io.apicurio.registry.config.RegistryConfigProperty)
+     */
+    @Override
+    public String get(RegistryConfigProperty property) {
+        return get(property, String.class);
+    }
+
+    /**
+     * @see io.apicurio.registry.config.RegistryConfigService#getOptional(io.apicurio.registry.config.RegistryConfigProperty)
+     */
+    @Override
+    public Optional<String> getOptional(RegistryConfigProperty property) {
+        return getOptional(property, String.class);
+    }
+
+    /**
+     * @see io.apicurio.registry.config.RegistryConfigService#get(io.apicurio.registry.config.RegistryConfigProperty, java.lang.Class)
+     */
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T get(RegistryConfigProperty property, Class<T> propertyType) {
+        // Possibly check the tenant property cache for a value.  If found return it.
+        T rval = property.isEditable() ? (T) tenantPropertyCache().get(property) : null;
+
+        // If not found, check for a global property.
+        if (rval == null) {
+            rval = (T) globalPropertyCache.computeIfAbsent(property, (key) -> {
+                Optional<? extends Object> value = ConfigProvider.getConfig().getOptionalValue(property.propertyName(), propertyType);
+                if (value.isPresent()) {
+                    return (T) value.get();
+                } else {
+                    return (T) property.defaultValue();
+                }
+            });
+        }
+
+        return rval;
+    }
+
+    /**
+     * @see io.apicurio.registry.config.RegistryConfigService#getOptional(io.apicurio.registry.config.RegistryConfigProperty, java.lang.Class)
+     */
+    @Override
+    public <T> Optional<T> getOptional(RegistryConfigProperty property, Class<T> propertyType) {
+        T rval = get(property, propertyType);
+        return Optional.ofNullable(rval);
+    }
+
+    public <T> T set(RegistryConfigProperty property, T newValue) {
+        storage.setConfigProperty(property.propertyName(), newValue);
+        return newValue;
+    }
+
+    private Map<RegistryConfigProperty, Object> tenantPropertyCache() {
+        String tenantId = tenantContext.getTenantIdOrElse(DEFAULT_TENANT_ID);
```

Comment:
actually if you just do `tenantContext.tenantId` you will get the same behaviour. tenant context will never return null tenantId

Reply:
Even better, thanks.